[Bug]: Nessie GC is not deleting files from S3 bucket after GC/delete command (expiry and orphan file clean up)

### What happened

Hi team,

We are trying to expire the old snapshots and delete the orphan files using the Nessie GC feature. Iceberg tables are created on top of s3 bucket, so after nessie gc we are expecting the expired files (data as well as metadata) to be deleted from the s3. However, the physical files are not deleted even after trying gc as well as explicit delete commands. Furthermore, we encountered following error while running nessie gc.

_Caused by: java.lang.RuntimeException: Failed to get paths from manifest file location.avro
        at org.projectnessie.gc.iceberg.IcebergContentToFiles.allDataAndDeleteFiles(IcebergContentToFiles.java:225)
        at org.projectnessie.gc.iceberg.IcebergContentToFiles.lambda$allManifestsAndDataFiles$2(IcebergContentToFiles.java:202)

Caused by: java.lang.IllegalArgumentException: Cannot parse partition spec fields, not an array: {"spec-id":0,"fields":[]}_

Can you please help us to understand this issue?  

### How to reproduce it

1. Create Iceberg table on s3 using Iceberg rest via Nessie (hosted via docker container)
2. Add some records in the table.
3. Get current snapshot id. 
4. Delete records from the table
5. Get current snapshot id. 
6. Run the following docker container to run the nessie GC. 
```
 nessie-gc:
    image: ghcr.io/projectnessie/nessie-gc:0.91.2
    ports:
      - "5435:5435"
    depends_on:
      nessie:
        condition: service_healthy
    command: gc --uri http://nessie:19120/api/v2 --iceberg=s3.access-key-id=${S3_KEY} --iceberg=s3.secret-access-key=${S3_SECRET} --jdbc --jdbc-url=${JDBC_URL} --jdbc-user=${JDBC_USERNAME} --jdbc-password=${JDBC_PASSWORD}
```
7. Check the gc log, docker logs container-id
8. Check the postgres tables if the references and live sets are deleted or not.
9. Check the s3 location if the expired or files marked as delete are deleted or not. 

                                                                                             

### Nessie server type (docker/uber-jar/built from source) and version

ghcr.io/projectnessie/nessie:0.90.4

### Client type (Ex: UI/Spark/pynessie ...) and version

_No response_

### Additional information

_No response_

Sure, PFB the exception log

2024-07-08 21:52:10,797 [ForkJoinPool-3-worker-4] INFO  org.apache.iceberg.CatalogUtil - Loading custom FileIO implementation: org.apache.iceberg.aws.s3.S3FileIO
java.lang.RuntimeException: java.lang.RuntimeException: Failed to get paths from manifest file s3://manifestfilelocation.avro
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:540)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:567)
        at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:653)
        at java.base/java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2822)
        at org.projectnessie.gc.expire.local.DefaultLocalExpire.expire(DefaultLocalExpire.java:73)
        at org.projectnessie.gc.tool.cli.commands.BaseRepositoryCommand.expire(BaseRepositoryCommand.java:250)
        at org.projectnessie.gc.tool.cli.commands.MarkAndSweep.call(MarkAndSweep.java:57)
        at org.projectnessie.gc.tool.cli.commands.BaseRepositoryCommand.call(BaseRepositoryCommand.java:84)
        at org.projectnessie.gc.tool.cli.commands.BaseCommand.call(BaseCommand.java:26)
        at org.projectnessie.gc.tool.cli.commands.BaseCommand.call(BaseCommand.java:21)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2045)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.projectnessie.gc.tool.cli.CLI.runMain(CLI.java:103)
        at org.projectnessie.gc.tool.cli.CLI.runMain(CLI.java:75)
        at org.projectnessie.gc.tool.cli.CLI.main(CLI.java:70)
Caused by: java.lang.RuntimeException: Failed to get paths from manifest file s3://manifestfilelocation.avro
        at org.projectnessie.gc.iceberg.IcebergContentToFiles.allDataAndDeleteFiles(IcebergContentToFiles.java:225)
        at org.projectnessie.gc.iceberg.IcebergContentToFiles.lambda$allManifestsAndDataFiles$2(IcebergContentToFiles.java:202)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
        at java.base/java.util.LinkedList$LLSpliterator.forEachRemaining(LinkedList.java:1249)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
        at java.base/java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:411)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
        at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
        at org.projectnessie.gc.contents.jdbc.JdbcHelper$ResultSetSplit.tryAdvance(JdbcHelper.java:143)
        at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:258)
        at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.count(ReferencePipeline.java:709)
        at org.projectnessie.gc.expire.PerContentDeleteExpired.identifyLiveFiles(PerContentDeleteExpired.java:136)
        at org.projectnessie.gc.expire.PerContentDeleteExpired.expire(PerContentDeleteExpired.java:73)
        at org.projectnessie.gc.expire.local.DefaultLocalExpire.expireSingleContent(DefaultLocalExpire.java:104)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
        at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
        at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
        at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.helpComplete(ForkJoinPool.java:1395)
        at java.base/java.util.concurrent.ForkJoinPool.helpComplete(ForkJoinPool.java:2104)
        at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:420)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:668)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:927)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:657)
        at org.projectnessie.gc.expire.local.DefaultLocalExpire.expireInForkJoinPool(DefaultLocalExpire.java:91)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1456)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.IllegalArgumentException: Cannot parse partition spec fields, not an array: {"spec-id":0,"fields":[]}
        at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:218)
        at org.apache.iceberg.PartitionSpecParser.buildFromJsonFields(PartitionSpecParser.java:123)
        at org.apache.iceberg.PartitionSpecParser.fromJsonFields(PartitionSpecParser.java:114)
        at org.apache.iceberg.PartitionSpecParser.lambda$fromJsonFields$4(PartitionSpecParser.java:119)
        at org.apache.iceberg.util.JsonUtil.parse(JsonUtil.java:98)
        at org.apache.iceberg.PartitionSpecParser.fromJsonFields(PartitionSpecParser.java:119)
        at org.apache.iceberg.ManifestReader.readPartitionSpec(ManifestReader.java:130)
        at org.apache.iceberg.ManifestReader.<init>(ManifestReader.java:114)
        at org.apache.iceberg.ManifestFiles.read(ManifestFiles.java:132)
        at org.apache.iceberg.ManifestFiles.open(ManifestFiles.java:244)
        at org.apache.iceberg.ManifestReaderUtil.readPathsFromManifest(ManifestReaderUtil.java:29)
        at org.projectnessie.gc.iceberg.IcebergContentToFiles.allDataAndDeleteFiles(IcebergContentToFiles.java:222)
        ... 57 more

--- a/gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java
+++ b/gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java
@@ -16,6 +16,7 @@
 package org.apache.iceberg;
 
 import com.google.common.collect.ImmutableList;
+import java.util.Map;
 import org.apache.iceberg.io.CloseableIterable;
 import org.apache.iceberg.io.FileIO;
 
@@ -23,10 +24,13 @@ public final class ManifestReaderUtil {
 
   private ManifestReaderUtil() {}
 
-  public static CloseableIterable<String> readPathsFromManifest(ManifestFile manifest, FileIO io) {
+  public static CloseableIterable<String> readPathsFromManifest(
+      ManifestFile manifest, Map<Integer, PartitionSpec> specsById, FileIO io) {
     // entry.file() is package private. Hence, this Util class under iceberg package.
     return CloseableIterable.transform(
-        ManifestFiles.open(manifest, io, null).select(ImmutableList.of("file_path")).liveEntries(),
+        ManifestFiles.open(manifest, io, specsById)
+            .select(ImmutableList.of("file_path"))
+            .liveEntries(),
         entry -> entry.file().path().toString());
   }
 }

--- a/gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/IcebergContentToFiles.java
+++ b/gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/IcebergContentToFiles.java
@@ -23,11 +23,13 @@ import com.google.errorprone.annotations.MustBeClosed;
 import jakarta.annotation.Nonnull;
 import java.io.IOException;
 import java.net.URI;
+import java.util.Map;
 import java.util.Objects;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 import org.apache.iceberg.ManifestFile;
 import org.apache.iceberg.ManifestReaderUtil;
+import org.apache.iceberg.PartitionSpec;
 import org.apache.iceberg.Snapshot;
 import org.apache.iceberg.TableMetadata;
 import org.apache.iceberg.TableMetadataParser;
@@ -153,6 +155,8 @@ public abstract class IcebergContentToFiles implements ContentToFiles {
     Snapshot snapshot =
         snapshotId < 0L ? tableMetadata.currentSnapshot() : tableMetadata.snapshot(snapshotId);
 
+    Map<Integer, PartitionSpec> specsById = tableMetadata.specsById();
+
     Stream<StorageUri> allFiles = elementaryUrisFromSnapshot(snapshot, contentReference);
 
     if (snapshot != null) {
@@ -166,7 +170,7 @@ public abstract class IcebergContentToFiles implements ContentToFiles {
                         try {
                           @SuppressWarnings("MustBeClosedChecker")
                           Stream<StorageUri> r =
-                              allManifestsAndDataFiles(io, snapshot, contentReference);
+                              allManifestsAndDataFiles(io, snapshot, specsById, contentReference);
                           return r;
                         } catch (Exception e) {
                           String msg =
@@ -193,25 +197,29 @@ public abstract class IcebergContentToFiles implements ContentToFiles {
 
   /**
    * For the given {@link Snapshot}, provide a {@link Stream} of all manifest files with {@link
-   * #allDataAndDeleteFiles(FileIO, ManifestFile, ContentReference) all included data and delete
-   * files}.
+   * #allDataAndDeleteFiles(FileIO, Map, ManifestFile, ContentReference) all included data and
+   * delete files}.
    */
   @MustBeClosed
   static Stream<StorageUri> allManifestsAndDataFiles(
-      FileIO io, Snapshot snapshot, ContentReference contentReference) {
-    return allManifests(io, snapshot)
+      FileIO io,
+      Snapshot snapshot,
+      Map<Integer, PartitionSpec> specsById,
+      ContentReference contentReference) {
+    return allManifests(io, specsById, snapshot)
         .flatMap(
             mf -> {
               StorageUri manifestFileLoc = manifestFileUri(mf, contentReference);
               @SuppressWarnings("MustBeClosedChecker")
               Stream<StorageUri> allDataAndDeleteFiles =
-                  allDataAndDeleteFiles(io, mf, contentReference);
+                  allDataAndDeleteFiles(io, specsById, mf, contentReference);
               return Stream.concat(Stream.of(manifestFileLoc), allDataAndDeleteFiles);
             });
   }
 
   /** Provide all {@link ManifestFile}s for the given {@link Snapshot}. */
-  static Stream<ManifestFile> allManifests(FileIO io, Snapshot snapshot) {
+  static Stream<ManifestFile> allManifests(
+      FileIO io, Map<Integer, PartitionSpec> specsById, Snapshot snapshot) {
     return snapshot.allManifests(io).stream();
   }
 
@@ -222,10 +230,13 @@ public abstract class IcebergContentToFiles implements ContentToFiles {
    */
   @MustBeClosed
   static Stream<StorageUri> allDataAndDeleteFiles(
-      FileIO io, ManifestFile manifestFile, ContentReference contentReference) {
+      FileIO io,
+      Map<Integer, PartitionSpec> specsById,
+      ManifestFile manifestFile,
+      ContentReference contentReference) {
     CloseableIterable<String> iter;
     try {
-      iter = ManifestReaderUtil.readPathsFromManifest(manifestFile, io);
+      iter = ManifestReaderUtil.readPathsFromManifest(manifestFile, specsById, io);
     } catch (Exception e) {
       throw new RuntimeException(
           "Failed to get paths from manifest file " + manifestFile.path(), e);

